type annotations

dimbleby opened the pull request:
It's kind of annoying that this library publishes a py.typed, but most of the API is not type-annotated.  Users who check their own code with mypy are obliged to scatter around `# type: ignore[no-untyped-call]` comments.

Possible points of interest:
- `joinpath` on the `SimplePath` seems to have been just wrong
- I've annotated some `Iterable`s as `Iterator`s, because (i) they are and (ii) `Distribution.from_name()` relies on this when it calls `next(..)` rather than `next(iter(..))`

This MR doesn't annotate the whole API: I've ducked the slightly difficult ones like `distributions()`, `select()` and `matches()`.  typeshed has annotations that are presumably satisfactory in practice, but it looks as though applying them here would be more invasive than I intend to be in this commit.  https://github.com/python/typeshed/blob/main/stdlib/importlib/metadata/__init__.pyi

This MR is a long way from annotating the whole project - `mypy --strict importlib_metadata` reports 171 errors, so it would take a bit more of a campaign to work through that.

jaraco commented:
FYI, @RonnyPfannschmidt was also considering tackling this effort (discussion in #importlib-metadata channel on PyPA discord).

dimbleby commented:
Cool.  I'm mostly just aiming for the low-hanging fruit here and perhaps testing the waters for receptiveness to further work.

The pipeline and I both are currently happy; I reckon I'm done fiddling with this one now.

RonnyPfannschmidt commented:
I have not yet started own work, great to see this start

jaraco reviewed (commented):
This looks good and nearly ready to merge. Just a few concerns to address.

dimbleby commented on importlib_metadata/__init__.py:
```diff
@@ -385,14 +393,14 @@ def from_name(cls, name: str):
-    def discover(cls, **kwargs):
-        """Return an iterable of Distribution objects for all packages.
+    def discover(cls, **kwargs) -> Iterator["Distribution"]:
+        """Return an iterator of Distribution objects for all packages.
 
         Pass a ``context`` or pass keyword arguments for constructing
         a context.
 
         :context: A ``DistributionFinder.Context`` object.
-        :return: Iterable of Distribution objects for all packages.
+        :return: Iterator of Distribution objects for all packages.
```
anything that's an `Iterator` is also an `Iterable`, returning an iterator is constraining for the implementation, not for the caller.

(which might still be a reason to reject this, if you think it's likely that you ever want to return something that's not actually an iterator).

jaraco replied:
I'm unsure what the best practice is here - to return an iterable or an iterator. I meant not to constrain alternate implementations. Regardless what's best, the currently-advertised return type is 'iterable', so let's retain that expectation and if you feel strongly about making it an iterator, let's address that concern separately.

dimbleby commented:
OK, I've pushed change that - I hope - err in the direction of avoiding and deferring any difficult questions

- prefer `Iterable` to `Iterator` (per the docstrings, and leaving other or future implementations a little more freedom)
  - as a result, added an `iter()` call in a place that was assuming it had an iterator
  - actually I notice now that I've annotated the `requires()` methods as returning `List` where the docstring used to say "iterator" (definitely wrong) and after my search-replace says "iterable" (true, but less specific than list).  Any preference on doing anything about that?

- completely ducked the typing of `Sectioned.valid()`, I don't like either of our suggestions but don't want to get stuck on that

- updated `StrPath` as suggested

- have `Distribution.at()` return `Distribution`s rather than committing to `PathDistribution`s
  - just spotted `MetadataPathFinder.find_distributions()` - is it ok for that to commit to `PathDistribution` or does that want changing too?

dimbleby commented on importlib_metadata/__init__.py:
```diff
@@ -403,7 +411,7 @@ def discover(cls, **kwargs):
 
     @staticmethod
-    def at(path):
+    def at(path: StrPath) -> "PathDistribution":
```
gone with `Distribution` as the unintrusive and uncommittal for now.  I see #445: you - or someone else - can always change this later.

jaraco commented:
> * just spotted `MetadataPathFinder.find_distributions()` - is it ok for that to commit to `PathDistribution` or does that want changing too?

Yes. `Distribution` is the right choice here. Thanks.

dimbleby commented:
> Yes. Distribution is the right choice here. Thanks.

the code that you have merged still used `PathDistribution`: https://github.com/python/importlib_metadata/blob/705a7571ec7c5abec4d4b008da3a58df7e5c94e7/importlib_metadata/__init__.py#L796-L798

jaraco commented:
> > Yes. Distribution is the right choice here. Thanks.
> 
> the code that you have merged still used `PathDistribution`:
> 

Aah. I was looking at the wrong thing (because it wasn't in the diff). I think `PathDistribution` is okay here because it's a specific implementation that does specifically only return `PathDistribution` (even though other `DistributionFinder`s might return a more generic `Distribution`). Thanks for flagging it.

jaraco commented:
I realized that the declaration of `Distribution.locate_file` is incorrect and have started work to correct it in #480.
